src/Compilers/CSharp/Test/Emit/PDB/PDBLambdaTests.cs

```diff
@@ -1381,5 +1381,283 @@ .maxstack  1
+
+            compilation.VerifyPdbLambdasAndClosures(source);
+        }
+
+        [Fact]
+        public void WithExpression_2()
+        {
+            var source = MarkedSource(WithWindowsLineBreaks(@"
+using System;
+record R(int X, int Y);
+
+class Test
+{
+    public static void M(int a)
+    <M><C:0>{
+        var x = new R(1, 2);
+        var b = 1;
+        var y = x with
+        {
+            X = new Func<int>(() => <L:0.0>a)(),
+            Y = new Func<int>(() => <L:1.0>b)()
+        };
+    }
+}
+"), removeTags: true); // We're validating offsets so need to remove tags entirely
+
+            // Use NetCoreApp in order to use records
+            var compilation = CreateCompilation(source.Tree, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.DebugDll);
+
+            compilation.VerifyPdbLambdasAndClosures(source);
+        }
+
+        [Fact]
+        public void WithExpression_3()
+        {
+            var source = MarkedSource(WithWindowsLineBreaks(@"
+using System;
+record R(int X, int Y);
+record Z(int A, R R);
+
+class Test
+{
+    public static void M(int a)
+    <M><C:0>{
+        var r = new R(1, 2);
+        var x = new Z(1, new R(2, 3));
+        var b = 1;
+        var y = x with
+        {
+            A = new Func<int>(() => <L:0.0>a)(),
+            R = r with
+            {
+                X = 4,
+                Y = new Func<int>(() => <L:1.0>b)()
+            }
+        };
+    }
+}
+"), removeTags: true); // We're validating offsets so need to remove tags entirely
+
+            // Use NetCoreApp in order to use records
+            var compilation = CreateCompilation(source.Tree, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.DebugDll);
+
+            compilation.VerifyPdbLambdasAndClosures(source);
+        }
+
+        [Fact]
+        public void Record_1()
+        {
+            var source = WithWindowsLineBreaks(@"
+using System;
+record D(int X)
+{
+    public int Y { get; set; } = new Func<int, int>(a => a + X).Invoke(1);
+}
+");
+
+            // Use NetCoreApp in order to use records
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.DebugDll);
+
+            compilation.VerifyPdb(@"
+    <symbols>
+      <files>
+        <file id=""1"" name="""" language=""C#"" />
+      </files>
+      <methods>
+        <method containingType=""D"" name="".ctor"" parameterNames=""X"">
+          <customDebugInfo>
+            <using>
+              <namespace usingCount=""1"" />
+            </using>
+            <encLocalSlotMap>
+              <slot kind=""30"" offset=""-1"" />
+            </encLocalSlotMap>
+            <encLambdaMap>
+              <methodOrdinal>0</methodOrdinal>
+              <closure offset=""-1"" />
+              <lambda offset=""-16"" closure=""0"" />
```

Comment:
Because of the negative offsets I couldn't use the markup style tests, so would appreciate someone smarter than me validating this.

Reply:
The negative offset is calculated like so: [http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Source/SourceMemberContainerSymbol.cs,1010](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Source/SourceMemberContainerSymbol.cs,1010)
I wouldn't expect any changes are needed for records since the properies defined by primary constructor syntax do not declare initializers.



---
In reply to: [595703676](https://github.com/dotnet/roslyn/pull/51926#discussion_r595703676) [](ancestors = 595703676)